detectron2/data/datasets/coco.py

```diff
@@ -280,6 +286,129 @@ def file2id(folder_path, file_path):
     return dataset_dicts
 
 
+def convert_to_coco_dict(dataset_name):
+    """
+    Convert a generic dataset into COCO json format
+
+    Generic dataset description can be found here:
+    https://detectron2.readthedocs.io/tutorials/datasets.html#register-a-dataset
+
+    COCO data format description can be found here:
+    http://cocodataset.org/#format-data
+
+    Args:
+        dataset_name: name of the source dataset
+    Returns:
+        coco_dict: serializable dict in COCO json format
+    """
+
+    dataset_dicts = DatasetCatalog.get(dataset_name)
+    categories = [
+        {"id": id, "name": name}
+        for id, name in enumerate(MetadataCatalog.get(dataset_name).thing_classes)
+    ]
+
+    logger.info("Converting dataset dicts into COCO format")
+    coco_images = []
+    coco_annotations = []
+
+    for image_dict in dataset_dicts:
+        coco_image = {
+            "id": image_dict["image_id"],
+            "width": image_dict["width"],
+            "height": image_dict["height"],
+            "file_name": image_dict["file_name"],
+        }
+        coco_images.append(coco_image)
+
+        anns_per_image = image_dict["annotations"]
+        for annotation in anns_per_image:
+            # create a new dict with only COCO fields
+            coco_annotation = {}
+
+            # COCO requirement: XYWH box format
+            bbox = annotation["bbox"]
+            bbox_mode = annotation["bbox_mode"]
+            bbox = BoxMode.convert(bbox, bbox_mode, BoxMode.XYWH_ABS)
+
+            # COCO requirement: instance area
+            if "segmentation" in annotation:
+                # Computing areas for instances by counting the pixels
+                segmentation = annotation["segmentation"]
+                # TODO: check segmentation type: RLE, BinaryMask or Polygon
+                polygons = PolygonMasks([segmentation])
+                area = polygons.area()[0].item()
+            else:
+                # Computing areas using bounding boxes
+                area = Boxes([bbox]).area()[0].item()
+
+            # COCO requirement:
+            #   linking annotations to images
+            #   "id" field must start with 1
+            coco_annotation["id"] = len(coco_annotations) + 1
+            coco_annotation["image_id"] = image_dict["image_id"]
+            coco_annotation["bbox"] = bbox
+            coco_annotation["area"] = area
+            coco_annotation["category_id"] = annotation["category_id"]
+            coco_annotation["iscrowd"] = annotation.get("iscrowd", 0)
+
+            # Add optional fields
+            if "keypoints" in annotation:
+                coco_annotation["keypoints"] = annotation["keypoints"]
```

Comment:
For keypoints support, I also had to add the following:
```
            if "num_keypoints" in annotation:
                coco_annotation["num_keypoints"] = annotation["num_keypoints"]
````
else it failed in cocoeval:

```
  File "venv/lib/python3.6/site-packages/pycocotools/cocoeval.py", line 111, in _prepare
    gt['ignore'] = (gt['num_keypoints'] == 0) or gt['ignore']
KeyError: 'num_keypoints'
```

Reply:
hmm, last time you could make a change proposal here on GH that I could just merge in directly.
Could you do this again? :D

Reply:
Done.

@ppwwyyxx exactly this is why I rather went with deepcopying all the fields, because the cocoapi wouldn't crash by it, and it could help with shoving custom data into detectron2

Reply:
But it still won't work for custom dataset that does not have this field, right?
Is this a redundant field that can be derived from keypoint visibility? Since it seems to be a required field of cocoapi (according to the error message), we should make sure it exists and derive it if necessary when creating the json. Just like how area is not a useful field for general training but is required by cocoapi.

Reply:
Done